### ISSUE TITLE
Allow inserting react elements as arguments to translate

### ISSUE THREAD

**Issue by neutrinog** [ryandrewjohnson/react-localize-redux#46]
It would be amazing if I could inject components into a string.

Let's suppose I have this translation
```json
{
  "greeting": "<div>Welcome ${user}!</div>"
}
```

I'd like to be able to do this:
```javascript

translate('greeting', { user: (
 <strong>Jon Doe</strong>
)});
```

The result would be:
```html
<div>Welcome <strong>Jon Doe</strong>!</div>
```

**Comment by ryandrewjohnson** [comment]
Yea I will have to give that some thought. Right now any HTML is rendered by using React's [dangerouslySetInnerHTML](https://zhenyong.github.io/react/tips/dangerously-set-inner-html.html) prop. 

In order to accomplish something like this you'd need something to convert a string version of JSX, and have that converted to fully rendered React component with props and state. I'm pretty sure there is no out of the box way to accomplish this at runtime. There may be a way to do it during a bundling step, and have it somehow run the translation string through the babel JSX transform for example. 

Anyways as I mentioned this will require some thought, but if you have any ideas on how this might be accomplished feel free to send my way.

**Comment by neutrinog** [comment]
Yes I agree this would be complicated. There is this library that converts HTML to a React tree https://github.com/aknuds1/html-to-react. However, this could quickly turn your tidy localize library into a monolith like some other localization libraries.

**Comment by neutrinog** [comment]
To make things a little simpler you could potentially expose your translation rendering so that a custom renderer could be used. Then a plugin could be made that uses https://github.com/aknuds1/html-to-react.

That way all the work is not on you to maintaining that complicated piece.

**Comment by thchia** [comment]
I have done a draft implementation of how I think this could be achieved without any major breaking changes*.

https://github.com/thchia/react-localize-redux/tree/data-component

The basic idea is that you can pass a React Component to `data` like so:

```javascript
const StrongName = ({ name }) => <strong>{name}</strong>
const translations = {
  greeting: 'Hello ${name}'
}
...

<Translate id='greeting' data={{ name: <StrongName name='Ryan' /> }} />
```

This is achieved by converting the `templater` function to return a string if there are no React components in `data` (like normal). If there _are_ React components, it returns an array representation of the string to be translated, like:

```javascript
[ 'Hello ', <StrongName /* ...props */ />]
```

Then in `getLocalizedElement`, if the result form `templater` is a string, perform the usual conversion to `dangerouslySetInnerHTML`. Otherwise, return

```javascript
React.createElement('span', null, ...arrayFromTemplater)
```

\* One problem I have not been able to solve is that if you want to use React components as arguments, you cannot have html in the translation. This is because `templater` will return something like this:

```javascript
['<p>Hello ', <Component />, '</p>']
```
Which will cannot be converted to `innerHTML` since it is a mixture of string literal HTML and React components.

Willing to discuss further if this looks like something you would consider supporting @ryandrewjohnson.

**Comment by ryandrewjohnson** [comment]
@thchia thanks for putting this together. Unfortunately I don't have much bandwidth to look into your solution, as what little time I do have right now is going towards getting v3 #67 ready. Once that is published I'll have some time to take a closer look at this. In the meantime maybe have a look at v3, as there are some significant changes that would most likely impact this implementation.

**Comment by thchia** [comment]
No worries, it's a fairly rough implementation anyway. In the meantime I will take a look at v3 and try to do a similar thing there.

**Comment by ryandrewjohnson** [comment]
@thchia I've officially added you as a collaborator. Thanks for all your help, and feel free to assign any issues you want to work on to yourself.

**Comment by neutrinog** [comment]
Just wanted to say it's exciting to see the potential for this feature!

**Comment by thchia** [comment]
Tracking this in #100
